pilot/pkg/bootstrap/server.go

```diff
@@ -591,29 +576,64 @@ func (s *Server) createK8sServiceControllers(serviceControllers *aggregate.Contr
 	return
 }
 
-// initMultiClusterController initializes multi cluster controller
-// currently implemented only for kubernetes registries
-func (s *Server) initMultiClusterController(args *PilotArgs) error {
-	if hasKubeRegistry(args) {
-		s.addStartFunc(func(stop <-chan struct{}) error {
-			secretController := clusterregistry.NewController(s.kubeClient,
-				args.Config.ClusterRegistriesNamespace,
-				s.clusterStore,
-				s.ServiceController,
-				s.EnvoyXdsServer,
-				args.Config.ControllerOptions.ResyncPeriod,
-				args.Config.ControllerOptions.WatchedNamespace,
-				args.Config.ControllerOptions.DomainSuffix)
-
-			// Start secret controller which watches for runtime secret Object changes and adds secrets dynamically
-			go secretController.Run(stop)
+// initMultiClusterSecretController initializes multi cluster secret controller
+// to watch for remote cluster additions or deletions.
+// Currently implemented only for kubernetes registries
+func (s *Server) initMultiClusterSecretController(args *PilotArgs) (err error) {
+	if checkForKubernetes(args.Service.Registries) {
+
+		err = secretcontroller.StartSecretController(s.kubeClient, s.addMemberCluster, s.deleteMemberCluster, args.Config.ClusterRegistriesNamespace)
+		if err != nil {
+			return fmt.Errorf("could not start secret controller: %v", err)
+		}
 
-			return nil
-		})
 	}
 	return nil
 }
 
+//addMemberCluster is passed to the secret controller as a callback to be called
+//when a remote cluster is added
+func (s *Server) addMemberCluster(clientset kubernetes.Interface, clusterID string) error {
+
+	stopCh := make(chan struct{})
+	kubectl := kube.NewController(clientset, kube.ControllerOptions{
+		WatchedNamespace: s.WatchedNamespace,
+		ResyncPeriod:     s.ResyncPeriod,
+		DomainSuffix:     s.DomainSuffix,
+	})
+	s.kubeRegistry[clusterID] = kubectl
+	s.ServiceController.AddRegistry(
+		aggregate.Registry{
+			Name:             serviceregistry.KubernetesRegistry,
+			ClusterID:        clusterID,
+			ServiceDiscovery: kubectl,
+			ServiceAccounts:  kubectl,
+			Controller:       kubectl,
+		})
+
+	kubectl.StopChan = stopCh
+
+	_ = kubectl.AppendServiceHandler(func(*model.Service, model.Event) { s.EnvoyXdsServer.ClearCacheFunc()() })
+	_ = kubectl.AppendInstanceHandler(func(*model.ServiceInstance, model.Event) { s.EnvoyXdsServer.ClearCacheFunc()() })
+	go kubectl.Run(stopCh)
+
+	return nil
+}
+
+// deleteMemberCluster is passed to the secret controller as a callback to be called
+// when a remote cluster is deleted
+func (s *Server) deleteMemberCluster(clusterID string) error {
+
+	s.ServiceController.DeleteRegistry(clusterID)
+	kubectl := s.kubeRegistry[clusterID]
+	close(kubectl.StopChan)
+	<-kubectl.StopChan
+	delete(s.kubeRegistry, clusterID)
+	s.EnvoyXdsServer.ClearCacheFunc()()
```

Comment:
what does this for?

Reply:
This is required to clear the cache so the services and endpoints from the remote cluster are cluster are removed.   This is the key line that is missing form the current code causing istio/istio#8554

Reply:
I doubt this really effect, when only secret delete(this maybe a mistake handly) but all controllers work.

Reply:
I tested this directly and proved that it does indeed fix the problem.   If you set up a multicluster environment and delete the secret associated with the remote cluster you can easily reproduce the issue.

Reply:
I double checked, it works.